Orm/Xtensive.Orm/Orm/EntitySetBase.cs

```diff
@@ -39,10 +39,45 @@ public abstract class EntitySetBase : SessionBound,
 #if DEBUG
     private static readonly string storageTestsAssemblyPrefix = "Xtensive.Orm.Tests";
 #endif
-    private static readonly object entitySetCachingRegion = new object();
     private static readonly Parameter<Tuple> keyParameter = new Parameter<Tuple>(WellKnown.KeyFieldName);
     internal static readonly Parameter<Entity> ownerParameter = new Parameter<Entity>("Owner");
 
+    private static readonly Func<FieldInfo, EntitySetBase, EntitySetTypeState> BuildEntitySetTypeState = (FieldInfo field, EntitySetBase entitySet) => {
+      var association = field.Associations.Last();
+      var query = association.UnderlyingIndex.GetQuery().Seek(context => context.GetValue(keyParameter));
+      var seek = entitySet.Session.Compile(query);
+      var ownerDescriptor = association.OwnerType.Key.TupleDescriptor;
+      var targetDescriptor = association.TargetType.Key.TupleDescriptor;
+
+      var itemColumnOffsets = association.AuxiliaryType == null
+        ? association.UnderlyingIndex.ValueColumns
+            .Where(ci => ci.IsPrimaryKey)
+            .Select(ci => ci.Field.MappingInfo.Offset)
+            .ToList()
+        : Enumerable.Range(0, targetDescriptor.Count).ToList();
+
+      var keyFieldCount = ownerDescriptor.Count + itemColumnOffsets.Count;
+      var keyFieldTypes = ownerDescriptor
+        .Concat(itemColumnOffsets.Select(i => targetDescriptor[i]))
+        .ToArray(keyFieldCount);
+      var keyDescriptor = TupleDescriptor.Create(keyFieldTypes);
+
+      var map = Enumerable.Range(0, ownerDescriptor.Count)
+        .Select(i => new Pair<int, int>(0, i))
+        .Concat(itemColumnOffsets.Select(i => new Pair<int, int>(1, i)))
+        .ToArray(keyFieldCount);
+      var seekTransform = new MapTransform(true, keyDescriptor, map);
+
+      Func<Tuple, Entity> itemCtor = null;
+      if (association.AuxiliaryType != null) {
+        itemCtor = DelegateHelper.CreateDelegate<Func<Tuple, Entity>>(null,
+          association.AuxiliaryType.UnderlyingType, DelegateHelper.AspectedFactoryMethodName,
+          Array.Empty<Type>());
+      }
+
+      return new EntitySetTypeState(seek, seekTransform, itemCtor, entitySet.GetItemCountQueryDelegate(field));
+    };
```

Comment:
I would not create such big-bodied ```Func``` and keep static method so it would be like so
```csharp
private static readonly Func<FieldInfo, EntitySetBase, EntitySetTypeState> EntitySetTypeStateFactory = BuildEntitySetTypeState;
```

Reply:
done